ako-infra/ingestion/vcf_k8s_controller.go

```diff
@@ -134,87 +136,89 @@ func (c *VCFK8sController) AddNCPSecretEventHandler(k8sinfo K8sinformers, stopCh
-	if !cache.WaitForCacheSync(stopCh, c.dynamicInformers.NSXNetworkConfigInformer.Informer().HasSynced) {
+	go c.informers.NSInformer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, c.informers.NSInformer.Informer().HasSynced) {
 		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
 	} else {
-		utils.AviLog.Info("Caches synced for nsxnetworkconfiguration informer")
+		utils.AviLog.Info("Caches synced for Namespace informer")
 	}
 }
 
-func addNSXNetworkConfigCount() error {
-	dynamicClient := lib.GetVCFDynamicClientSet()
-	crdClient := dynamicClient.Resource(lib.NSXNetworkConfigGVR)
-	crdList, err := crdClient.List(context.TODO(), metav1.ListOptions{})
+func (c *VCFK8sController) addWorkloadNamespaceCount() error {
+	nsList, err := c.informers.NSInformer.Lister().List(labels.Set(nil).AsSelector())
 	if err != nil {
 		return err
 	}
 
-	NsxNetworkConfigCount = len(crdList.Items)
+	count := getWorkloadNamespaceCount(nsList)
+	WorkloadNamespaceCount = count
 	return nil
 }
 
-func handleNSXNetworkConfigAdd() {
-	dynamicClient := lib.GetVCFDynamicClientSet()
-	crdClient := dynamicClient.Resource(lib.NSXNetworkConfigGVR)
-	crdList, err := crdClient.List(context.TODO(), metav1.ListOptions{})
+func (c *VCFK8sController) handleNamespaceAdd() {
+	countLock.Lock()
+	defer countLock.Unlock()
+	nsList, err := c.informers.NSInformer.Lister().List(labels.Set(nil).AsSelector())
 	if err != nil {
 		utils.AviLog.Error(err)
 		return
 	}
 
+	count := getWorkloadNamespaceCount(nsList)
+
 	// Only when before the addition, the count was 0, (and now it becomes more than 0),
 	// we must reconfigure the SEG, by rebooting AKO. On reboot AKO ensures SEG configuration.
-	if NsxNetworkConfigCount == 0 && len(crdList.Items) > 0 {
-		utils.AviLog.Fatalf("First NSXNetworkConfigurations added in cluster. Rebooting AKO for infra configuration.")
+	if WorkloadNamespaceCount == 0 && count > 0 {
+		utils.AviLog.Fatalf("First Workload Namespace added in cluster. Rebooting AKO for infra configuration.")
 	}
-	NsxNetworkConfigCount = len(crdList.Items)
+	WorkloadNamespaceCount = count
 }
 
-func handleNSXNetworkConfigDelete() {
-	dynamicClient := lib.GetVCFDynamicClientSet()
-	crdClient := dynamicClient.Resource(lib.NSXNetworkConfigGVR)
-	crdList, err := crdClient.List(context.TODO(), metav1.ListOptions{})
+func (c *VCFK8sController) handleNamespaceDelete() {
+	countLock.Lock()
+	nsList, err := c.informers.NSInformer.Lister().List(labels.Set(nil).AsSelector())
 	if err != nil {
 		utils.AviLog.Error(err)
 		return
 	}
 
-	NsxNetworkConfigCount = len(crdList.Items)
-	if len(crdList.Items) > 0 {
-		utils.AviLog.Infof("%d NSXNetworkConfigurations exist in the cluster. Skipping deconfiguration.", len(crdList.Items))
+	count := getWorkloadNamespaceCount(nsList)
+	WorkloadNamespaceCount = count
+	if count > 0 {
+		utils.AviLog.Infof("%d Workload Namespace exist in the cluster. Skipping deconfiguration.", count)
 		return
 	}
+	countLock.Unlock()
```

Comment:
@chauhanshubham : I think "unlock()" call over here will cause deadlock. If `count > 0` , we will return from line 217 and unlock() will not be called. We need defer unlock() on function exit.

Reply:
yes! thanks.